File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/RideHistoryRepository.kt
package expo.modules.vescapecore.telemetry

import android.content.Context
import androidx.room.withTransaction

private const val RIDE_BUCKET_BATCH_SIZE = 100
private const val MAX_RIDE_PAGE_SIZE = 50
private val RIDE_BREAK_BOUNDARIES = setOf("disconnected", "app_stop", "error")

internal data class RideRoutePoint(val latitude: Double, val longitude: Double)

internal data class RideSessionAggregate(
  val deviceId: String,
  var deviceName: String,
  var boundaryBefore: String,
  var firstBucketStartMs: Long,
  var startAtMs: Long,
  var endAtMs: Long,
  val blockIds: MutableList<String>,
  var blockCount: Int,
  var sampleCount: Int,
  var gpsPointCount: Int,
  var preciseGpsPointCount: Int,
  var avgSpeedSampleCount: Int,
  var avgSpeedWeightedSum: Double,
  var movingStartAtMs: Long?,
  var movingEndAtMs: Long?,
  var distanceDeltaM: Double,
  var distanceDeltaCount: Int,
  var gpsDistanceM: Double,
  var gpsDistanceCount: Int,
  var topSpeedKmh: Double,
  var maxTempMosfet: Double?,
  var maxTempMotor: Double?,
  var maxDuty: Double,
  var batteryUsedWh: Double,
  var batteryRegenWh: Double,
  var firstLatitude: Double?,
  var firstLongitude: Double?,
  var latitudeSum: Double,
  var longitudeSum: Double,
  var coordinateCount: Int,
  var minLatitude: Double?,
  var maxLatitude: Double?,
  var minLongitude: Double?,
  var maxLongitude: Double?,
  var faultCount: Int,
  val routePoints: MutableList<RideRoutePoint>,
) {
  val distanceM: Double?
    get() = when {
      distanceDeltaCount > 0 -> distanceDeltaM
      gpsDistanceCount > 0 -> gpsDistanceM
      else -> null
    }
}

/** Complete-ride paging over minute buckets. JS never observes a provisional, cut-off ride. */
// @parity /modules/vescape-core/ios/telemetry/RideHistoryRepository.swift
internal class RideHistoryRepository private constructor(private val context: Context) {
  private val database = TelemetryDatabase.get(context)
  private val dao = database.telemetryDao()

  /** @parity /modules/vescape-core/src/index.ts `RideHistoryPage` */
  suspend fun getPage(options: Map<String, Any?>): Map<String, Any?> {
    val limit = ((options["limit"] as? Number)?.toInt() ?: 10).coerceIn(1, MAX_RIDE_PAGE_SIZE)
    val gapMs = rideSplitGapMs()
    return database.withTransaction {
      var beforeExclusive = (options["cursorBeforeMs"] as? Number)?.toLong() ?: Long.MAX_VALUE
      val buckets = mutableListOf<TelemetryMinuteBucketEntity>()
      var hasOlderBuckets = true
      var complete = emptyList<RideSessionAggregate>()

      while (hasOlderBuckets && complete.size < limit) {
        val beforeInclusive = if (beforeExclusive == Long.MAX_VALUE) Long.MAX_VALUE else beforeExclusive - 1L
        val fetched = dao.getHistoryBuckets(0, Long.MAX_VALUE, beforeInclusive, null, RIDE_BUCKET_BATCH_SIZE + 1)
        val batch = fetched.take(RIDE_BUCKET_BATCH_SIZE)
        if (batch.isEmpty()) {
          hasOlderBuckets = false
          break
        }
        buckets.addAll(batch)
        beforeExclusive = batch.minOf { it.bucketStartMs }
        hasOlderBuckets = fetched.size > RIDE_BUCKET_BATCH_SIZE
        val markerFrom = buckets.minOf { it.firstSampleAtMs } - gapMs
        val markerTo = buckets.maxOf { it.lastSampleAtMs } + TELEMETRY_BUCKET_SIZE_MS
        val markers = dao.getMarkers(markerFrom, markerTo, null)
        val grouped = groupRideSessions(buckets, markers, gapMs).filter { it.avgSpeedSampleCount > 0 }
        complete = completeRideSessions(grouped, hasOlderBuckets)
      }

      val sorted = complete.sortedByDescending { it.startAtMs }
      val cutoff = sorted.getOrNull(limit - 1)?.firstBucketStartMs
      val page = if (cutoff == null) sorted else sorted.filter { it.firstBucketStartMs >= cutoff }
      val hasMore = hasOlderBuckets || (cutoff != null && sorted.any { it.firstBucketStartMs < cutoff })
      mapOf(
        "sessions" to page.map(::rideSessionMap),
        "hasMore" to hasMore,
        "nextCursorBeforeMs" to if (hasMore) page.lastOrNull()?.firstBucketStartMs else null,
      )
    }
  }

  private suspend fun rideSplitGapMs(): Long {
    val minutes = AppDataRepository.get(context).getSettings()["rideSplitGapMinutes"] as? Number
    return (minutes?.toLong() ?: DEFAULT_RIDE_SPLIT_GAP_MINUTES.toLong()) * 60_000L
  }

  companion object {
    @Volatile private var instance: RideHistoryRepository? = null

    fun get(context: Context): RideHistoryRepository = instance ?: synchronized(this) {
      instance ?: RideHistoryRepository(context.applicationContext).also { instance = it }
    }

    fun resetForDatabaseSwap() = synchronized(this) { instance = null }
  }
}

/**
 * Buckets arrive newest-first, so the only ride that may still grow backwards is the OLDEST one in
 * the window. Dropping the newest instead hides the ride being recorded right now.
 * @parity /modules/vescape-core/ios/telemetry/RideHistoryRepository.swift `completeRideSessions`
 */
internal fun completeRideSessions(
  grouped: List<RideSessionAggregate>,
  hasOlderBuckets: Boolean,
): List<RideSessionAggregate> = if (hasOlderBuckets) grouped.drop(1) else grouped

/**
 * Native source of Ride boundaries and aggregates used by both History pages and Profile stats.
 * @parity /modules/vescape-core/ios/telemetry/RideHistoryRepository.swift `groupRideSessions`
 */
internal fun groupRideSessions(
  buckets: List<TelemetryMinuteBucketEntity>,
  markers: List<TelemetryMarkerEntity>,
  gapMs: Long,
): List<RideSessionAggregate> {
  val sessions = mutableListOf<RideSessionAggregate>()
  var current: RideSessionAggregate? = null
  var previous: TelemetryMinuteBucketEntity? = null

  for (bucket in buckets.sortedBy { it.firstSampleAtMs }) {
    if (bucket.sampleCount <= 0) continue
    val boundary = rideBoundaryForBucket(bucket, markers)
    val split = current == null || current.deviceId != bucket.deviceId ||
      (previous != null && bucket.firstSampleAtMs - previous.lastSampleAtMs > gapMs) ||
      RIDE_BREAK_BOUNDARIES.contains(boundary)
    if (split) {
      current?.let(sessions::add)
      current = newRideAggregate(bucket, boundary)
    }
    mergeRideBucket(current ?: continue, bucket)
    previous = bucket
  }
  current?.let(sessions::add)
  return sessions
}

private fun newRideAggregate(bucket: TelemetryMinuteBucketEntity, boundary: String) = RideSessionAggregate(
  deviceId = bucket.deviceId,
  deviceName = bucket.deviceName ?: UNKNOWN_TELEMETRY_DEVICE_NAME,
  boundaryBefore = boundary,
  firstBucketStartMs = bucket.bucketStartMs,
  startAtMs = bucket.firstSampleAtMs,
  endAtMs = bucket.lastSampleAtMs,
  blockIds = mutableListOf(), blockCount = 0, sampleCount = 0, gpsPointCount = 0,
  preciseGpsPointCount = 0, avgSpeedSampleCount = 0, avgSpeedWeightedSum = 0.0,
  movingStartAtMs = null, movingEndAtMs = null, distanceDeltaM = 0.0,
  distanceDeltaCount = 0, gpsDistanceM = 0.0, gpsDistanceCount = 0, topSpeedKmh = 0.0,
  maxTempMosfet = null, maxTempMotor = null, maxDuty = 0.0, batteryUsedWh = 0.0,
  batteryRegenWh = 0.0, firstLatitude = null, firstLongitude = null, latitudeSum = 0.0,
  longitudeSum = 0.0, coordinateCount = 0, minLatitude = null, maxLatitude = null,
  minLongitude = null, maxLongitude = null, faultCount = 0, routePoints = mutableListOf(),
)

private fun mergeRideBucket(session: RideSessionAggregate, bucket: TelemetryMinuteBucketEntity) {
  session.firstBucketStartMs = minOf(session.firstBucketStartMs, bucket.bucketStartMs)
  session.startAtMs = minOf(session.startAtMs, bucket.firstSampleAtMs)
  session.endAtMs = maxOf(session.endAtMs, bucket.lastSampleAtMs)
  session.blockIds.add("${bucket.deviceId}:${bucket.bucketStartMs}")
  session.blockCount++
  session.sampleCount += bucket.sampleCount
  session.gpsPointCount += bucket.gpsPointCount
  session.preciseGpsPointCount += bucket.preciseGpsPointCount
  val movingCount = bucket.movingSpeedSampleCount ?: bucket.sampleCount
  session.avgSpeedSampleCount += movingCount
  session.avgSpeedWeightedSum += if (bucket.movingSpeedSampleCount != null) {
    (bucket.sumMovingAbsSpeedCentiKmh ?: 0L) / 100.0
  } else bucket.sumAbsSpeedCentiKmh / 100.0
  bucket.firstMovingAtMs?.let { session.movingStartAtMs = session.movingStartAtMs?.let { old -> minOf(old, it) } ?: it }
  bucket.lastMovingAtMs?.let { session.movingEndAtMs = session.movingEndAtMs?.let { old -> maxOf(old, it) } ?: it }
  val odometerDistance = rideDistanceDeltaM(bucket)
  if (odometerDistance != null) { session.distanceDeltaM += odometerDistance; session.distanceDeltaCount++ }
  if (bucket.gpsDistanceCm > 0) { session.gpsDistanceM += bucket.gpsDistanceCm / 100.0; session.gpsDistanceCount++ }
  session.topSpeedKmh = maxOf(session.topSpeedKmh, bucket.maxAbsSpeedCentiKmh / 100.0)
  bucket.maxTempMosfetDeciC?.let { value -> session.maxTempMosfet = maxOf(session.maxTempMosfet ?: value / 10.0, value / 10.0) }
  bucket.maxTempMotorDeciC?.let { value -> session.maxTempMotor = maxOf(session.maxTempMotor ?: value / 10.0, value / 10.0) }
  session.maxDuty = maxOf(session.maxDuty, bucket.maxDutyAbsPermille / 1000.0)
  session.batteryUsedWh += bucket.batteryUsedWhMilli / 1000.0
  session.batteryRegenWh += bucket.batteryRegenWhMilli / 1000.0
  session.faultCount += bucket.faultCount
  if (bucket.firstLatitudeE7 != null && bucket.firstLongitudeE7 != null) {
    val latitude = bucket.firstLatitudeE7 / 1e7
    val longitude = bucket.firstLongitudeE7 / 1e7
    if (session.firstLatitude == null) { session.firstLatitude = latitude; session.firstLongitude = longitude }
    session.latitudeSum += latitude; session.longitudeSum += longitude; session.coordinateCount++
    session.minLatitude = minOf(session.minLatitude ?: latitude, latitude)
    session.maxLatitude = maxOf(session.maxLatitude ?: latitude, latitude)
    session.minLongitude = minOf(session.minLongitude ?: longitude, longitude)
    session.maxLongitude = maxOf(session.maxLongitude ?: longitude, longitude)
    session.routePoints.add(RideRoutePoint(latitude, longitude))
  }
}

private fun rideBoundaryForBucket(bucket: TelemetryMinuteBucketEntity, markers: List<TelemetryMarkerEntity>): String =
  markers.lastOrNull { marker ->
    marker.occurredAtMs >= bucket.firstSampleAtMs - 5_000L &&
      marker.occurredAtMs <= bucket.firstSampleAtMs + 1_000L &&
      (marker.deviceId ?: UNKNOWN_TELEMETRY_DEVICE_ID) == bucket.deviceId
  }?.type ?: "none"

private fun rideDistanceDeltaM(bucket: TelemetryMinuteBucketEntity): Double? {
  val first = bucket.firstOdometerCm ?: return null
  val last = bucket.lastOdometerCm ?: return null
  return (last - first).coerceAtLeast(0L) / 100.0
}

/** @parity /modules/vescape-core/src/index.ts `RideHistorySession` */
internal fun rideSessionMap(session: RideSessionAggregate): Map<String, Any?> {
  val avgSpeed = if (session.avgSpeedSampleCount > 0) session.avgSpeedWeightedSum / session.avgSpeedSampleCount else 0.0
  return mapOf(
    "id" to "${session.deviceId.ifBlank { "unknown" }}:${session.startAtMs}:${session.endAtMs}",
    "deviceId" to session.deviceId.ifBlank { null }, "deviceName" to session.deviceName,
    "startAtMs" to session.startAtMs, "endAtMs" to session.endAtMs,
    "movingStartAtMs" to session.movingStartAtMs, "movingEndAtMs" to session.movingEndAtMs,
    "blockIds" to session.blockIds, "blockCount" to session.blockCount,
    "sampleCount" to session.sampleCount, "gpsPointCount" to session.gpsPointCount,
    "preciseGpsPointCount" to session.preciseGpsPointCount, "distanceM" to session.distanceM,
    "maxSpeedKmh" to session.topSpeedKmh, "avgSpeedKmh" to avgSpeed,
    "maxTempMosfet" to session.maxTempMosfet, "maxTempMotor" to session.maxTempMotor,
    "maxDuty" to session.maxDuty, "batteryUsedWh" to session.batteryUsedWh,
    "batteryRegenWh" to session.batteryRegenWh, "firstLatitude" to session.firstLatitude,
    "firstLongitude" to session.firstLongitude,
    "centerLatitude" to if (session.coordinateCount > 0) session.latitudeSum / session.coordinateCount else null,
    "centerLongitude" to if (session.coordinateCount > 0) session.longitudeSum / session.coordinateCount else null,
    "minLatitude" to session.minLatitude, "maxLatitude" to session.maxLatitude,
    "minLongitude" to session.minLongitude, "maxLongitude" to session.maxLongitude,
    "faultCount" to session.faultCount, "boundaryBefore" to session.boundaryBefore,
    "routePoints" to session.routePoints.map { mapOf("latitude" to it.latitude, "longitude" to it.longitude) },
  )
}

File: modules/vescape-core/ios/telemetry/RideHistoryRepository.swift
import Foundation
import GRDB

internal func historyBoardNames(_ db: Database) throws -> [String: String] {
  try Row.fetchAll(db, sql: "SELECT id, name FROM boards").reduce(into: [String: String]()) {
    $0[$1["id"] as String] = $1["name"] as String
  }
}

internal func historyBatteryConfigs(_ db: Database) throws -> [String: [String: Any]] {
  try Row.fetchAll(
    db,
    sql: "SELECT board_id, value_json FROM board_settings WHERE key = 'batteryConfig'"
  ).reduce(into: [String: [String: Any]]()) { result, row in
    let data = Data((row["value_json"] as String).utf8)
    if let config = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
      result[row["board_id"] as String] = config
    }
  }
}

private let rideBucketBatchSize = 100
private let maxRidePageSize = 50
private let rideBreakBoundaries: Set<String> = ["disconnected", "app_stop", "error"]

/// Keep every recording and Board in the boundary minute; the bucket limit is a soft limit.
/// Both queries run in the caller's read transaction so the cursor and lookahead share a snapshot.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/TelemetryDao.kt `getRideBuckets`
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/TelemetryDao.kt `hasRideBucketsBefore`
internal func fetchRideHistoryBucketBatch(
  _ db: Database,
  beforeMs: Int64,
  limit: Int = rideBucketBatchSize
) throws -> (buckets: [Row], hasOlder: Bool) {
  let buckets = try Row.fetchAll(
    db,
    sql: """
      SELECT * FROM telemetry_minute_buckets
      WHERE bucket_start_ms < ? AND bucket_start_ms >= (
        SELECT MIN(bucket_start_ms) FROM (
          SELECT bucket_start_ms FROM telemetry_minute_buckets
          WHERE bucket_start_ms < ? ORDER BY bucket_start_ms DESC LIMIT ?
        )
      )
      ORDER BY bucket_start_ms DESC
      """,
    arguments: [beforeMs, beforeMs, limit]
  )
  guard let oldest = buckets.last?["bucket_start_ms"] as Int64? else { return ([], false) }
  let hasOlder = try Bool.fetchOne(
    db,
    sql: "SELECT EXISTS(SELECT 1 FROM telemetry_minute_buckets WHERE bucket_start_ms < ?)",
    arguments: [oldest]
  ) ?? false
  return (buckets, hasOlder)
}

internal struct RideRoutePoint {
  let latitude: Double
  let longitude: Double
}

internal struct RideSessionAggregate {
  /// Owning Board (`boards.id`), empty when the buckets match no saved Board (ADR 0028).
  let boardId: String
  /// Owning Ride Recording, or `LEGACY_RIDE_RECORDING_ID` for buckets with no durable identity.
  /// A real recording *is* the history entry: gap length and disconnect markers never split it.
  let recordingId: String
  var boundaryBefore: String
  var firstBucketStartMs: Int64
  var startAtMs: Int64
  var endAtMs: Int64
  var blockIds: [String] = []
  var blockCount = 0
  var sampleCount = 0
  var gpsPointCount = 0
  var preciseGpsPointCount = 0
  var avgSpeedSampleCount = 0
  var avgSpeedWeightedSum = 0.0
  var movingStartAtMs: Int64?
  var movingEndAtMs: Int64?
  var distanceDeltaM = 0.0
  var distanceDeltaCount = 0
  var gpsDistanceM = 0.0
  var gpsDistanceCount = 0
  var topSpeedKmh = 0.0
  var maxTempMosfet: Double?
  var maxTempMotor: Double?
  var maxDuty = 0.0
  var batteryUsedWh = 0.0
  var batteryRegenWh = 0.0
  var firstLatitude: Double?
  var firstLongitude: Double?
  var latitudeSum = 0.0
  var longitudeSum = 0.0
  var coordinateCount = 0
  var minLatitude: Double?
  var maxLatitude: Double?
  var minLongitude: Double?
  var maxLongitude: Double?
  var routePoints: [RideRoutePoint] = []

  var distanceM: Double? {
    if distanceDeltaCount > 0 { return distanceDeltaM }
    if gpsDistanceCount > 0 { return gpsDistanceM }
    return nil
  }
}

/// Complete-ride paging over minute buckets. JS never observes a provisional, cut-off ride.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/RideHistoryRepository.kt
internal final class RideHistoryRepository {
  static let shared = RideHistoryRepository()
  private let poolProvider: () throws -> DatabasePool
  private let gapMsProvider: () throws -> Int64

  internal init(
    poolProvider: @escaping () throws -> DatabasePool = { try TelemetryDatabase.requirePool() },
    gapMsProvider: @escaping () throws -> Int64 = {
      let minutes = telemetryInt(try AppDataRepository.shared.getSettings()["rideSplitGapMinutes"] ?? nil)
      return Int64(minutes ?? DEFAULT_RIDE_SPLIT_GAP_MINUTES) * 60_000
    }
  ) {
    self.poolProvider = poolProvider
    self.gapMsProvider = gapMsProvider
  }

  /// @parity /modules/vescape-core/src/index.ts `RideHistoryPage`
  func getPage(_ options: [String: Any]) throws -> [String: Any?] {
    let limit = min(maxRidePageSize, max(1, telemetryInt(options["limit"]) ?? 10))
    var beforeMs = telemetryLong(options["cursorBeforeMs"]) ?? Int64.max
    let gapMs = try gapMsProvider()
    let pool = try poolProvider()
    return try pool.read { db in
      // Names resolve from `boards` on read, never off the bucket row (ADR 0028).
      let boardNames = try historyBoardNames(db)
      var buckets: [Row] = []
      var complete: [RideSessionAggregate] = []
      var hasOlderBuckets = true
      while hasOlderBuckets && complete.count < limit {
        let batch = try fetchRideHistoryBucketBatch(db, beforeMs: beforeMs)
        if batch.buckets.isEmpty { hasOlderBuckets = false; break }
        buckets.append(contentsOf: batch.buckets)
        beforeMs = batch.buckets.map { $0["bucket_start_ms"] as Int64 }.min() ?? beforeMs
        hasOlderBuckets = batch.hasOlder
        let markerFrom = (buckets.map { $0["first_sample_at_ms"] as Int64 }.min() ?? 0) - gapMs
        let markerTo = (buckets.map { $0["last_sample_at_ms"] as Int64 }.max() ?? 0) + TELEMETRY_BUCKET_SIZE_MS
        let markers = try Row.fetchAll(
          db,
          sql: "SELECT * FROM telemetry_markers WHERE occurred_at_ms >= ? AND occurred_at_ms <= ? ORDER BY occurred_at_ms ASC",
          arguments: [markerFrom, markerTo]
        )
        let grouped = groupRideSessions(buckets: buckets, markers: markers, gapMs: gapMs)
          .filter { $0.avgSpeedSampleCount > 0 }
        complete = completeRideSessions(grouped, hasOlderBuckets: hasOlderBuckets)
      }
      let sorted = complete.sorted { $0.startAtMs > $1.startAtMs }
      let cutoff = sorted.indices.contains(limit - 1) ? sorted[limit - 1].firstBucketStartMs : nil
      let page = cutoff.map { value in sorted.filter { $0.firstBucketStartMs >= value } } ?? sorted
      let hasMore = hasOlderBuckets || cutoff.map { value in sorted.contains { $0.firstBucketStartMs < value } } == true
      return [
        "sessions": page.map { rideSessionMap($0, boardNames: boardNames) },
        "hasMore": hasMore,
        "nextCursorBeforeMs": hasMore ? page.last?.firstBucketStartMs : nil,
      ]
    }
  }

}

/// Buckets arrive newest-first, so the only ride that may still grow backwards is the OLDEST one in
/// the window. Dropping the newest instead hides the ride being recorded right now.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/RideHistoryRepository.kt `completeRideSessions`
internal func completeRideSessions(
  _ grouped: [RideSessionAggregate],
  hasOlderBuckets: Bool
) -> [RideSessionAggregate] {
  hasOlderBuckets ? Array(grouped.dropFirst()) : grouped
}

/// Native Ride boundaries/aggregates shared by History pages and Profile stats.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/RideHistoryRepository.kt `groupRideSessions`
internal func groupRideSessions(buckets: [Row], markers: [Row], gapMs: Int64) -> [RideSessionAggregate] {
  var sessions: [RideSessionAggregate] = []
  var current: RideSessionAggregate?
  var previous: Row?
  for bucket in buckets.sorted(by: { ($0["first_sample_at_ms"] as Int64) < ($1["first_sample_at_ms"] as Int64) }) {
    let boundary = rideBoundaryForBucket(bucket, markers: markers)
    let boardId = bucket["board_id"] as String
    let recordingId = bucket["recording_id"] as String
    // Legacy rows have no recording identity, so they keep reconstructing rides from gaps and
    // break markers. A row that carries one needs neither: the recording *is* the entry, and a
    // dropout of any length inside it stays one ride (ADR 0038).
    let legacy = recordingId == LEGACY_RIDE_RECORDING_ID
    let split = current == nil || current?.boardId != boardId || current?.recordingId != recordingId ||
      (legacy && (previous.map { (bucket["first_sample_at_ms"] as Int64) - ($0["last_sample_at_ms"] as Int64) > gapMs } ?? false)) ||
      (legacy && rideBreakBoundaries.contains(boundary))
    if split {
      if let current { sessions.append(current) }
      current = RideSessionAggregate(
        boardId: boardId,
        recordingId: recordingId,
        boundaryBefore: boundary,
        firstBucketStartMs: bucket["bucket_start_ms"] as Int64,
        startAtMs: bucket["first_sample_at_ms"] as Int64,
        endAtMs: bucket["last_sample_at_ms"] as Int64
      )
    }
    if var aggregate = current { mergeRideBucket(bucket, into: &aggregate); current = aggregate }
    previous = bucket
  }
  if let current { sessions.append(current) }
  return sessions
}

private func mergeRideBucket(_ bucket: Row, into session: inout RideSessionAggregate) {
  let bucketStart = bucket["bucket_start_ms"] as Int64
  session.firstBucketStartMs = min(session.firstBucketStartMs, bucketStart)
  session.startAtMs = min(session.startAtMs, bucket["first_sample_at_ms"] as Int64)
  session.endAtMs = max(session.endAtMs, bucket["last_sample_at_ms"] as Int64)
  // Same identity `getHistory` gives a bucket: two recordings of one Board can share a minute.
  session.blockIds.append("\(session.boardId):\(session.recordingId):\(bucketStart)")
  session.blockCount += 1
  session.sampleCount += bucket["sample_count"] as Int
  session.gpsPointCount += bucket["gps_point_count"] as Int
  session.preciseGpsPointCount += bucket["precise_gps_point_count"] as Int
  if let movingCount = bucket["moving_speed_sample_count"] as Int? {
    session.avgSpeedSampleCount += movingCount
    session.avgSpeedWeightedSum += Double((bucket["sum_moving_abs_speed_centi_kmh"] as Int64?) ?? 0) / 100.0
  } else {
    session.avgSpeedSampleCount += bucket["sample_count"] as Int
    session.avgSpeedWeightedSum += Double(bucket["sum_abs_speed_centi_kmh"] as Int64) / 100.0
  }
  if let value = bucket["first_moving_at_ms"] as Int64? { session.movingStartAtMs = session.movingStartAtMs.map { min($0, value) } ?? value }
  if let value = bucket["last_moving_at_ms"] as Int64? { session.movingEndAtMs = session.movingEndAtMs.map { max($0, value) } ?? value }
  if let distance = rideDistanceDeltaM(bucket) { session.distanceDeltaM += distance; session.distanceDeltaCount += 1 }
  let gpsDistance = Double(bucket["gps_distance_cm"] as Int64) / 100.0
  if gpsDistance > 0 { session.gpsDistanceM += gpsDistance; session.gpsDistanceCount += 1 }
  session.topSpeedKmh = max(session.topSpeedKmh, Double(bucket["max_abs_speed_centi_kmh"] as Int) / 100.0)
  if let value = bucket["max_temp_mosfet_deci_c"] as Int? { session.maxTempMosfet = max(session.maxTempMosfet ?? Double(value) / 10.0, Double(value) / 10.0) }
  if let value = bucket["max_temp_motor_deci_c"] as Int? { session.maxTempMotor = max(session.maxTempMotor ?? Double(value) / 10.0, Double(value) / 10.0) }
  session.maxDuty = max(session.maxDuty, Double(bucket["max_duty_abs_permille"] as Int) / 1000.0)
  session.batteryUsedWh += Double(bucket["battery_used_wh_milli"] as Int64) / 1000.0
  session.batteryRegenWh += Double(bucket["battery_regen_wh_milli"] as Int64) / 1000.0
  if let latE7 = bucket["first_latitude_e7"] as Int64?, let lonE7 = bucket["first_longitude_e7"] as Int64? {
    let latitude = Double(latE7) / 1e7, longitude = Double(lonE7) / 1e7
    if session.firstLatitude == nil { session.firstLatitude = latitude; session.firstLongitude = longitude }
    session.latitudeSum += latitude; session.longitudeSum += longitude; session.coordinateCount += 1
    session.minLatitude = min(session.minLatitude ?? latitude, latitude); session.maxLatitude = max(session.maxLatitude ?? latitude, latitude)
    session.minLongitude = min(session.minLongitude ?? longitude, longitude); session.maxLongitude = max(session.maxLongitude ?? longitude, longitude)
    session.routePoints.append(RideRoutePoint(latitude: latitude, longitude: longitude))
  }
}

private func rideBoundaryForBucket(_ bucket: Row, markers: [Row]) -> String {
  markers.last { marker in
    let occurred = marker["occurred_at_ms"] as Int64
    return occurred >= (bucket["first_sample_at_ms"] as Int64) - 5_000 &&
      occurred <= (bucket["first_sample_at_ms"] as Int64) + 1_000 &&
      ((marker["board_id"] as String?) ?? "") == (bucket["board_id"] as String)
  }.map { $0["type"] as String } ?? "none"
}

private func rideDistanceDeltaM(_ bucket: Row) -> Double? {
  guard let first = bucket["first_odometer_cm"] as Int64?, let last = bucket["last_odometer_cm"] as Int64? else { return nil }
  return Double(max(0, last - first)) / 100.0
}

/// @parity /modules/vescape-core/src/index.ts `RideHistorySession`
internal func rideSessionMap(_ session: RideSessionAggregate, boardNames: [String: String]) -> [String: Any?] {
  let average = session.avgSpeedSampleCount > 0 ? session.avgSpeedWeightedSum / Double(session.avgSpeedSampleCount) : 0
  return [
    "id": session.recordingId.isEmpty ? "\(session.boardId.isEmpty ? "unknown" : session.boardId):\(session.startAtMs):\(session.endAtMs)" : session.recordingId,
    "recordingId": session.recordingId.isEmpty ? nil : session.recordingId,
    "boardId": session.boardId.isEmpty ? nil : session.boardId,
    "boardName": boardNames[session.boardId] ?? UNKNOWN_TELEMETRY_BOARD_NAME,
    "startAtMs": session.startAtMs, "endAtMs": session.endAtMs, "movingStartAtMs": session.movingStartAtMs,
    "movingEndAtMs": session.movingEndAtMs, "blockIds": session.blockIds, "blockCount": session.blockCount,
    "sampleCount": session.sampleCount, "gpsPointCount": session.gpsPointCount,
    "preciseGpsPointCount": session.preciseGpsPointCount, "distanceM": session.distanceM,
    "maxSpeedKmh": session.topSpeedKmh, "avgSpeedKmh": average, "maxTempMosfet": session.maxTempMosfet,
    "maxTempMotor": session.maxTempMotor, "maxDuty": session.maxDuty, "batteryUsedWh": session.batteryUsedWh,
    "batteryRegenWh": session.batteryRegenWh, "firstLatitude": session.firstLatitude,
    "firstLongitude": session.firstLongitude,
    "centerLatitude": session.coordinateCount > 0 ? session.latitudeSum / Double(session.coordinateCount) : nil,
    "centerLongitude": session.coordinateCount > 0 ? session.longitudeSum / Double(session.coordinateCount) : nil,
    "minLatitude": session.minLatitude, "maxLatitude": session.maxLatitude, "minLongitude": session.minLongitude,
    "maxLongitude": session.maxLongitude, "boundaryBefore": session.boundaryBefore,
    "routePoints": session.routePoints.map { ["latitude": $0.latitude, "longitude": $0.longitude] },
  ]
}
